Drop support for DBAL 2, PHP 7.2, Symfony 3, Persistence 1

greg0ire opened the pull request:
DBAL 2 will no longer be supported tomorrow, and DBAL 3 requires PHP 7.3

greg0ire committed:
Drop support for DBAL 2 and PHP 7.2

DBAL 3 requires PHP 7.3

greg0ire commented on lib/Doctrine/Migrations/Tools/TransactionHelper.php:
```diff
@@ -72,11 +72,9 @@ private static function inTransaction(Connection $connection): bool
+        } catch (LogicException $e) {
         }
 
         $innermostConnection = $connection;
```
Todo: check if the code below is still useful

greg0ire replied:
From what I understood, it will be useful until DBAL 4, when we are sure that `getNativeConnection` can no longer throw a `LogicException`.

greg0ire commented on .appveyor.yml:
```diff
@@ -13,7 +13,6 @@ cache:
 
 environment:
   matrix:
-    - php: 7.2.25
```
Todo: determine what our policy for testing on appveyor shall be

greg0ire commented:
:thinking: just noticed we apparently still support unmaintained versions of Symfony, we should bump that as well.

greg0ire committed:
Drop unmaintained version of Symfony libs

greg0ire committed:
Drop unmaintained version of doctrine/persistence

derrabus commented on composer.json:
```diff
@@ -24,32 +24,32 @@
     ],
     "require": {
-        "php": "^7.2 || ^8.0",
+        "php": "^7.3 || ^8.0",
```
Shall we drop 7.3 as well?

greg0ire replied:
I don't see a reason to, but also, I don't see a reason not to, so sure, let's do it!

derrabus commented on composer.json:
```diff
@@ -24,32 +24,32 @@
+        "php": "^7.3 || ^8.0",
         "composer-runtime-api": "^2",
-        "doctrine/dbal": "^2.11 || ^3.0",
+        "doctrine/dbal": "^3.0",
```
We need to require `^3.3` here because you've dropped some compat code that we would still need for DBAL 3.2 and below, like checking for `getNativeConnection()`.

derrabus commented on composer.json:
```diff
@@ -24,32 +24,32 @@
-        "symfony/cache": "^3.4.26 || ^4.2.12 || ^5.0 || ^6.0",
-        "symfony/process": "^3.4 || ^4.0 || ^5.0 || ^6.0",
-        "symfony/yaml": "^3.4 || ^4.0 || ^5.0 || ^6.0"
+        "phpunit/phpunit": "^9.4",
```
```suggestion
        "phpunit/phpunit": "^9.5",
```

PHPUnit 9.4 is dead and buried.

greg0ire commented on phpcs.xml.dist:
```diff
@@ -17,6 +17,18 @@
+
+    <rule ref="SlevomatCodingStandard.TypeHints.PropertyTypeHint.MissingNativeTypeHint">
+        <exclude-pattern>*/lib/*</exclude-pattern>
+    </rule>
```
Maybe I should remove those, `git add --patch` the parts that are backwards compatible, and restore these lines. But that would probably be best in a separate PR for the sake of review.

greg0ire replied:
Just tried it, and it's not that complicated to untangle, I'll push it as an extra commit.

greg0ire committed:
Use typed properties where possible

This includes:
- private properties
- public properties of final classes

protected properties must have the same type or lack thereof as their
parent.

greg0ire commented:
I think that the 2 issues reported by the BC check can be safely ignored.
